### PR TITLE
Skale 1538

### PR DESCRIPTION
**Pull request by kladkogex**
Four node tests passed.

Added several new tests to consensust and Travis.

### CODE REVIEW

File: .idea/mapreduce_consensus.iml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <module classpath="CMake" type="CPP_MODULE" version="4">
   <component name="FacetManager">
     <facet type="Python" name="Python facet">
```
Comment (DimaStebaev) [inline .idea/mapreduce_consensus.iml:4]: Add .idea folder to .gitignore and remove from the repo.

File: Consensust.cpp [hunk 2]
```diff
@@ -29,157 +29,121 @@
 #include "Log.h"
 #include "node/ConsensusEngine.h"
 
+#include "time.h"
```
Comment (DimaStebaev) [inline Consensust.cpp:32]: It is better to use std::chrono

File: Consensust.cpp [hunk 3]
```diff
@@ -29,157 +29,121 @@
...
-
-
-
-void Consensust::testInit() {
-
-    setConfigDirPath(boost::filesystem::system_complete("."));
+class StartFromScratch {
+public:
+    StartFromScratch() {
+        int i = system("rm -rf /tmp/*.db");
```
Comment (DimaStebaev) [inline Consensust.cpp:44]: Never use `rm -f` in scripts

File: Consensust.cpp [hunk 4]
```diff
@@ -29,157 +29,121 @@
...
+        i++; // make compiler happy
```
Comment (DimaStebaev) [inline Consensust.cpp:45]: Compiler is not happy: variable `i` is never used.

File: Consensust.h [hunk 5]
```diff
@@ -23,6 +23,10 @@
 
 #pragma once
 
+
+#define DEFAULT_RUNNING_TIME_MS 30000
```
Comment (DimaStebaev) [inline Consensust.h:27]: Use constants instead of macros.

File: Log.cpp [hunk 6]
```diff
@@ -36,144 +36,163 @@ using namespace std;
...
     }
+
+    assert(logger);
+
     return logger;
 }
 
-void Log::setGlobalLogLevel( string& _s ) {
-    globalLogLevel = logLevelFromString( _s );
+void Log::setGlobalLogLevel(string &_s) {
```
Comment (DimaStebaev) [inline Log.cpp:82]: const reference can be used here.

File: Log.cpp [hunk 7]
```diff
@@ -36,144 +36,163 @@ using namespace std;
...
-    for ( auto&& item : loggers ) {
-        item.second->set_level( globalLogLevel );
+    for (auto &&item : loggers) {
+        item.second->set_level(globalLogLevel);
     }
 }
 
-void Log::setConfigLogLevel( string& _s ) {
-    auto configLogLevel = logLevelFromString( _s );
+void Log::setConfigLogLevel(string &_s) {
```
Comment (DimaStebaev) [inline Log.cpp:90]: const reference can be used here

File: Log.cpp [hunk 8]
```diff
@@ -36,144 +36,163 @@ using namespace std;
...
-    Log::configLogger->set_level( configLogLevel );
+    Log::configLogger->set_level(configLogLevel);
 }
 
 
-level_enum Log::logLevelFromString( string& _s ) {
-    for ( int i = 0; i < 7; i++ ) {
-        if ( _s == level_names[i] ) {
-            return level_enum( i );
+level_enum Log::logLevelFromString(string &_s) {
```
Comment (DimaStebaev) [inline Log.cpp:97]: const reference can be used here

File: Log.cpp [hunk 9]
```diff
@@ -36,144 +36,163 @@ using namespace std;
...
         }
     }
 
 
-    throw ParsingException( "Unknown level name " + _s , __CLASS_NAME__);
+    BOOST_THROW_EXCEPTION(ParsingException("Unknown level name " + _s, __CLASS_NAME__));
 }
 
-shared_ptr< spdlog::logger > Log::loggerForClass( const char* _s ) {
+shared_ptr<spdlog::logger> Log::loggerForClass(const char *_s) {
```
Comment (DimaStebaev) [inline Log.cpp:108]: It is better to use std::string

File: Makefile [hunk 10]
```diff
@@ -1,5 +1,5 @@
 # CMAKE generated file: DO NOT EDIT!
```
Comment (DimaStebaev) [inline Makefile:1]: This file is generated by cmake. Remove it from the repo.

File: abstracttcpserver/AbstractServerAgent.h [hunk 11]
```diff
@@ -41,7 +41,7 @@ class AbstractServerAgent : public Agent {
 
     const string name;
 
-    ptr<thread> networkReadThread;
+    thread* networkReadThread;
```
Comment (DimaStebaev) [inline abstracttcpserver/AbstractServerAgent.h:44]: Raw pointer is used.

File: catchup/server/CatchupServerAgent.h [hunk 12]
```diff
@@ -44,20 +44,29 @@
...
 public:
     CatchupServerAgent(Schain &_schain, ptr<TCPServerSocket> _s);
     ~CatchupServerAgent() override;
 
     CatchupWorkerThreadPool *getCatchupWorkerThreadPool() const;
 
-    ptr<vector<uint8_t>> createCatchupResponseHeader(ptr<Connection> _connectionEnvelope,
-                                nlohmann::json _jsonRequest, ptr<CatchupResponseHeader> _responseHeader);
+    ptr<vector<uint8_t>> createResponseHeaderAndBinary(ptr<Connection> _connectionEnvelope,
+                                                       nlohmann::json _jsonRequest, ptr<Header>& _responseHeader);
```
Comment (DimaStebaev) [inline catchup/server/CatchupServerAgent.h:69]: There is no sense to pass pointer by reference

File: chains/Schain.h [hunk 13]
```diff
@@ -104,115 +95,92 @@ class Schain : public Agent {
...
 
     ptr<IO> io;
 
+    ptr<ReceivedBlockProposalsDatabase> blockProposalsDatabase;
 
+    ptr<ReceivedSigSharesDatabase> blockSigSharesDatabase;
 
-    Node& node;
-
+    Node* node;
```
Comment (DimaStebaev) [inline chains/Schain.h:118]: Raw pointer is used.

File: datastructures/BlockProposalSet.cpp [hunk 14]
```diff
@@ -36,79 +36,79 @@
...
 
-    if ( proposals.count( _index) == 0 ) {
-//        LOG(trace,
-//            "Proposal did not yet arrive. Total proposals:" + to_string(proposals.size()));
+    if ( proposals.count((uint64_t) _index) == 0 ) {
         return nullptr;
     }
 
-    return proposals.at(_index);
+    return proposals.at((uint64_t)_index);
```
Comment (DimaStebaev) [inline datastructures/BlockProposalSet.cpp:111]: Don't use c style casts.

File: Consensust.cpp [hunk 15]
```diff
@@ -29,157 +29,121 @@
...
 
+TEST_CASE_METHOD(StartFromScratch, "Run basic consensus", "[consensus-basic]") {
 
-    engine.exitGracefully();
+    engine = new ConsensusEngine();
+    engine->parseConfigsAndCreateAllNodes(Consensust::getConfigDirPath());
+    engine->slowStartBootStrapTest();
+    usleep(1000 * Consensust::getRunningTimeMS()); /* Flawfinder: ignore */
 
+    assert(engine->nodesCount() > 0);
```
Comment (DimaStebaev) [inline Consensust.cpp:104]: Do not use nonconst methods in assert statements.

File: Consensust.cpp [hunk 16]
```diff
@@ -29,157 +29,121 @@
...
+    assert(engine->getLargestCommittedBlockID() > 0);
```
Comment (DimaStebaev) [inline Consensust.cpp:105]: The same.

File: blockfinalize/client/BlockFinalizeDownloader.cpp [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+    CHECK_ARGUMENT(responseHeader > 0);
+
+    auto fragmentSize = readFragmentSize(responseHeader);
+    auto blockSize = readBlockSize(responseHeader);
+    auto blockHash = readBlockHash(responseHeader);
+
+    auto serializedFragment = make_shared<vector<uint8_t> >(fragmentSize);
+
+    try {
+        getSchain()->getIo()->readBytes(_socket->getDescriptor(), (in_buffer *) serializedFragment->data(),
```
Comment (DimaStebaev) [inline blockfinalize/client/BlockFinalizeDownloader.cpp:228]: Do not use c-style pointer casts.

**Review by sergiy-skalelabs — APPROVED** [review]

File: blockfinalize/client/BlockFinalizeDownloader.cpp (hunk 17 above)
Comment (kladkogex) [reply blockfinalize/client/BlockFinalizeDownloader.cpp:228]: Ok - will fix

**Pull request merged**